File: go/shield/channel.go
package shield

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/binary"
	"errors"
	"io"
	"net"
)

// Protocol constants
const (
	protocolVersion  = 1
	maxMessageSize   = 16 * 1024 * 1024 // 16 MB
	maxHandshakeSize = 1024
)

// Handshake message types
const (
	handshakeClientHello = 1
	handshakeServerHello = 2
	handshakeFinished    = 3
)

// Channel errors
var (
	ErrInvalidHandshake  = errors.New("shield: invalid handshake")
	ErrUnsupportedVersion = errors.New("shield: unsupported protocol version")
	ErrMessageTooLarge   = errors.New("shield: message too large")
	ErrConnectionClosed  = errors.New("shield: connection closed")
)

// ChannelConfig holds channel configuration.
type ChannelConfig struct {
	Password           string
	Service            string
	Iterations         int
	HandshakeTimeoutMs int64
}

// NewChannelConfig creates a new channel configuration.
func NewChannelConfig(password, service string) *ChannelConfig {
	return &ChannelConfig{
		Password:           password,
		Service:            service,
		Iterations:         200000,
		HandshakeTimeoutMs: 30000,
	}
}

// WithIterations sets custom PBKDF2 iterations.
func (c *ChannelConfig) WithIterations(iterations int) *ChannelConfig {
	c.Iterations = iterations
	return c
}

// WithTimeout sets handshake timeout.
func (c *ChannelConfig) WithTimeout(timeoutMs int64) *ChannelConfig {
	c.HandshakeTimeoutMs = timeoutMs
	return c
}

// ShieldChannel provides encrypted bidirectional communication.
type ShieldChannel struct {
	conn    io.ReadWriteCloser
	session *RatchetSession
	service string
}

// Connect initiates a client connection with PAKE handshake.
func Connect(conn io.ReadWriteCloser, config *ChannelConfig) (*ShieldChannel, error) {
	ch := &ShieldChannel{conn: conn, service: config.Service}

	// Step 1: Generate client salt and send ClientHello
	clientSalt := make([]byte, 16)
	if _, err := rand.Read(clientSalt); err != nil {
		return nil, err
	}
	if err := ch.sendHandshake(handshakeClientHello, clientSalt); err != nil {
		return nil, err
	}

	// Step 2: Receive ServerHello
	serverHello, err := ch.recvHandshake(handshakeServerHello)
	if err != nil {
		return nil, err
	}
	if len(serverHello) != 48 {
		return nil, ErrInvalidHandshake
	}

	finalSalt := serverHello[:16]
	serverContribution := serverHello[16:48]

	// Step 3: Derive our contribution and send it
	clientContribution := PAKEDerive(config.Password, finalSalt, "client", config.Iterations)
	if err := ch.sendHandshake(handshakeFinished, clientContribution); err != nil {
		return nil, err
	}

	// Compute session key
	sessionKey := ch.computeSessionKey(config, finalSalt, clientContribution, serverContribution)

	// Create ratchet session
	ch.session, err = NewRatchetSession(sessionKey, true)
	if err != nil {
		return nil, err
	}

	// Exchange confirmations
	if err := ch.sendConfirmation(sessionKey, true); err != nil {
		return nil, err
	}
	if err := ch.verifyConfirmation(sessionKey, false); err != nil {
		return nil, err
	}

	return ch, nil
}

// Accept waits for a client connection with PAKE handshake.
func Accept(conn io.ReadWriteCloser, config *ChannelConfig) (*ShieldChannel, error) {
	ch := &ShieldChannel{conn: conn, service: config.Service}

	// Step 1: Receive ClientHello
	clientHello, err := ch.recvHandshake(handshakeClientHello)
	if err != nil {
		return nil, err
	}
	if len(clientHello) != 16 {
		return nil, ErrInvalidHandshake
	}

	// Mix salts
	serverSalt := make([]byte, 16)
	if _, err := rand.Read(serverSalt); err != nil {
		return nil, err
	}
	finalSalt := make([]byte, 16)
	for i := 0; i < 16; i++ {
		finalSalt[i] = serverSalt[i] ^ clientHello[i]
	}

	// Derive server contribution
	serverContribution := PAKEDerive(config.Password, finalSalt, "server", config.Iterations)

	// Step 2: Send ServerHello
	serverHello := make([]byte, 48)
	copy(serverHello[:16], finalSalt)
	copy(serverHello[16:], serverContribution)
	if err := ch.sendHandshake(handshakeServerHello, serverHello); err != nil {
		return nil, err
	}

	// Step 3: Receive client contribution
	clientFinished, err := ch.recvHandshake(handshakeFinished)
	if err != nil {
		return nil, err
	}
	if len(clientFinished) != 32 {
		return nil, ErrInvalidHandshake
	}

	// Compute session key
	sessionKey := ch.computeSessionKey(config, finalSalt, serverContribution, clientFinished)

	// Create ratchet session
	ch.session, err = NewRatchetSession(sessionKey, false)
	if err != nil {
		return nil, err
	}

	// Exchange confirmations
	if err := ch.verifyConfirmation(sessionKey, true); err != nil {
		return nil, err
	}
	if err := ch.sendConfirmation(sessionKey, false); err != nil {
		return nil, err
	}

	return ch, nil
}

// Send encrypts and sends a message.
func (ch *ShieldChannel) Send(data []byte) error {
	if len(data) > maxMessageSize {
		return ErrMessageTooLarge
	}

	encrypted, err := ch.session.Encrypt(data)
	if err != nil {
		return err
	}

	return ch.writeFrame(encrypted)
}

// Recv receives and decrypts a message.
func (ch *ShieldChannel) Recv() ([]byte, error) {
	encrypted, err := ch.readFrame()
	if err != nil {
		return nil, err
	}

	return ch.session.Decrypt(encrypted)
}

// Service returns the service identifier.
func (ch *ShieldChannel) Service() string {
	return ch.service
}

// MessagesSent returns the send message count.
func (ch *ShieldChannel) MessagesSent() uint64 {
	return ch.session.SendCounter()
}

// MessagesReceived returns the receive message count.
func (ch *ShieldChannel) MessagesReceived() uint64 {
	return ch.session.RecvCounter()
}

// Close closes the channel.
func (ch *ShieldChannel) Close() error {
	return ch.conn.Close()
}

// --- Internal helpers ---

func (ch *ShieldChannel) computeSessionKey(config *ChannelConfig, salt, localContribution, remoteContribution []byte) []byte {
	baseKey := PAKECombine(localContribution, remoteContribution)
	passwordKey := PAKEDerive(config.Password, salt, "session", config.Iterations)

	combined := make([]byte, 64)
	copy(combined[:32], baseKey)
	copy(combined[32:], passwordKey)

	h := sha256.Sum256(combined)
	return h[:]
}

func (ch *ShieldChannel) sendHandshake(msgType byte, data []byte) error {
	frame := make([]byte, 4+len(data))
	frame[0] = protocolVersion
	frame[1] = msgType
	binary.BigEndian.PutUint16(frame[2:4], uint16(len(data)))
	copy(frame[4:], data)

	_, err := ch.conn.Write(frame)
	return err
}

func (ch *ShieldChannel) recvHandshake(expectedType byte) ([]byte, error) {
	header := make([]byte, 4)
	if _, err := io.ReadFull(ch.conn, header); err != nil {
		return nil, err
	}

	if header[0] != protocolVersion {
		return nil, ErrUnsupportedVersion
	}

	if header[1] != expectedType {
		return nil, ErrInvalidHandshake
	}

	length := binary.BigEndian.Uint16(header[2:4])
	if length > maxHandshakeSize {
		return nil, ErrInvalidHandshake
	}

	data := make([]byte, length)
	if _, err := io.ReadFull(ch.conn, data); err != nil {
		return nil, err
	}

	return data, nil
}

func (ch *ShieldChannel) sendConfirmation(sessionKey []byte, isClient bool) error {
	label := "server-confirm"
	if isClient {
		label = "client-confirm"
	}

	mac := hmac.New(sha256.New, sessionKey)
	mac.Write([]byte(label))
	confirm := mac.Sum(nil)[:16]

	return ch.writeFrame(confirm)
}

func (ch *ShieldChannel) verifyConfirmation(sessionKey []byte, expectClient bool) error {
	received, err := ch.readFrame()
	if err != nil {
		return err
	}
	if len(received) != 16 {
		return ErrInvalidHandshake
	}

	label := "server-confirm"
	if expectClient {
		label = "client-confirm"
	}

	mac := hmac.New(sha256.New, sessionKey)
	mac.Write([]byte(label))
	expected := mac.Sum(nil)[:16]

	if subtle.ConstantTimeCompare(received, expected) != 1 {
		return ErrAuthenticationFailed
	}

	return nil
}

func (ch *ShieldChannel) writeFrame(data []byte) error {
	header := make([]byte, 4)
	binary.BigEndian.PutUint32(header, uint32(len(data)))

	if _, err := ch.conn.Write(header); err != nil {
		return err
	}
	_, err := ch.conn.Write(data)
	return err
}

func (ch *ShieldChannel) readFrame() ([]byte, error) {
	header := make([]byte, 4)
	if _, err := io.ReadFull(ch.conn, header); err != nil {
		return nil, err
	}

	length := binary.BigEndian.Uint32(header)
	if length > maxMessageSize {
		return nil, ErrMessageTooLarge
	}

	data := make([]byte, length)
	if _, err := io.ReadFull(ch.conn, data); err != nil {
		return nil, err
	}

	return data, nil
}

// ShieldListener listens for multiple Shield connections.
type ShieldListener struct {
	listener net.Listener
	config   *ChannelConfig
}

// NewShieldListener creates a new Shield listener.
func NewShieldListener(listener net.Listener, config *ChannelConfig) *ShieldListener {
	return &ShieldListener{
		listener: listener,
		config:   config,
	}
}

// Accept accepts the next connection.
func (l *ShieldListener) Accept() (*ShieldChannel, error) {
	conn, err := l.listener.Accept()
	if err != nil {
		return nil, err
	}

	return Accept(conn, l.config)
}

// Config returns the configuration.
func (l *ShieldListener) Config() *ChannelConfig {
	return l.config
}

// Close closes the listener.
func (l *ShieldListener) Close() error {
	return l.listener.Close()
}

File: python/shield/channel.py
"""
Shield Secure Channel - TLS/SSH-like secure transport using symmetric crypto.

Provides encrypted bidirectional communication with:
- PAKE-based handshake (no certificates needed)
- Forward secrecy via key ratcheting
- Message authentication and replay protection

Example:
    >>> from shield.channel import ShieldChannel, ChannelConfig
    >>>
    >>> # Both parties share a password
    >>> config = ChannelConfig("shared-secret", "my-service")
    >>>
    >>> # Server side
    >>> server = ShieldChannel.accept(conn, config)
    >>>
    >>> # Client side
    >>> client = ShieldChannel.connect(conn, config)
    >>>
    >>> # Send/receive with forward secrecy
    >>> client.send(b"Hello server!")
    >>> msg = server.recv()
"""

import hmac
import hashlib
import struct
import secrets
from typing import Optional, BinaryIO, Union
from dataclasses import dataclass

from shield.exchange import PAKEExchange
from shield.ratchet import RatchetSession


# Protocol constants
PROTOCOL_VERSION = 1
MAX_MESSAGE_SIZE = 16 * 1024 * 1024  # 16 MB

# Handshake message types
HANDSHAKE_CLIENT_HELLO = 1
HANDSHAKE_SERVER_HELLO = 2
HANDSHAKE_FINISHED = 3


@dataclass
class ChannelConfig:
    """Channel configuration."""
    password: str
    service: str
    iterations: int = 200000
    handshake_timeout_ms: int = 30000

    def with_iterations(self, iterations: int) -> 'ChannelConfig':
        """Set custom PBKDF2 iterations."""
        self.iterations = iterations
        return self

    def with_timeout(self, timeout_ms: int) -> 'ChannelConfig':
        """Set handshake timeout."""
        self.handshake_timeout_ms = timeout_ms
        return self


class ShieldChannel:
    """
    Shield secure channel for encrypted communication.

    Provides TLS-like security using only symmetric cryptography:
    - PAKE handshake establishes shared key from password
    - RatchetSession provides forward secrecy
    - All messages authenticated with HMAC
    """

    def __init__(self, stream: BinaryIO, session: RatchetSession, service: str):
        self._stream = stream
        self._session = session
        self._service = service

    @classmethod
    def connect(cls, stream: BinaryIO, config: ChannelConfig) -> 'ShieldChannel':
        """
        Connect as client (initiator).

        Performs PAKE handshake and establishes encrypted channel.

        Args:
            stream: Underlying transport (socket, etc.)
            config: Channel configuration with shared password

        Returns:
            Connected ShieldChannel
        """
        # Step 1: Generate client salt and send ClientHello
        client_salt = secrets.token_bytes(16)
        cls._send_handshake(stream, HANDSHAKE_CLIENT_HELLO, client_salt)

        # Step 2: Receive ServerHello (server's salt + contribution)
        server_hello = cls._recv_handshake(stream, HANDSHAKE_SERVER_HELLO)
        if len(server_hello) != 48:
            raise ValueError("Invalid ServerHello")

        # Extract final salt and server contribution
        final_salt = server_hello[:16]
        server_contribution = server_hello[16:48]

        # Step 3: Derive our contribution and send it
        client_contribution = PAKEExchange.derive(
            config.password, final_salt, "client", config.iterations
        )
        cls._send_handshake(stream, HANDSHAKE_FINISHED, client_contribution)

        # Compute session key
        session_key = cls._compute_session_key(
            config, final_salt, client_contribution, server_contribution
        )

        # Create ratchet session (client is initiator)
        session = RatchetSession(session_key, is_initiator=True)

        # Exchange confirmation messages
        cls._send_confirmation(stream, session_key, is_client=True)
        cls._verify_confirmation(stream, session_key, expect_client=False)

        return cls(stream, session, config.service)

    @classmethod
    def accept(cls, stream: BinaryIO, config: ChannelConfig) -> 'ShieldChannel':
        """
        Accept connection as server.

        Waits for client handshake and establishes encrypted channel.

        Args:
            stream: Underlying transport (socket, etc.)
            config: Channel configuration with shared password

        Returns:
            Connected ShieldChannel
        """
        # Step 1: Receive ClientHello (client's proposed salt)
        client_hello = cls._recv_handshake(stream, HANDSHAKE_CLIENT_HELLO)
        if len(client_hello) != 16:
            raise ValueError("Invalid ClientHello")

        # Mix client salt with server salt for freshness
        server_salt = secrets.token_bytes(16)
        final_salt = bytes(a ^ b for a, b in zip(server_salt, client_hello))

        # Derive server contribution
        server_contribution = PAKEExchange.derive(
            config.password, final_salt, "server", config.iterations
        )

        # Step 2: Send ServerHello (final salt + server contribution)
        server_hello = final_salt + server_contribution
        cls._send_handshake(stream, HANDSHAKE_SERVER_HELLO, server_hello)

        # Step 3: Receive client contribution
        client_finished = cls._recv_handshake(stream, HANDSHAKE_FINISHED)
        if len(client_finished) != 32:
            raise ValueError("Invalid Finished")

        client_contribution = client_finished

        # Compute session key
        session_key = cls._compute_session_key(
            config, final_salt, server_contribution, client_contribution
        )

        # Create ratchet session (server is not initiator)
        session = RatchetSession(session_key, is_initiator=False)

        # Exchange confirmation messages
        cls._verify_confirmation(stream, session_key, expect_client=True)
        cls._send_confirmation(stream, session_key, is_client=False)

        return cls(stream, session, config.service)

    def send(self, data: bytes) -> None:
        """
        Send encrypted message.

        Message is encrypted with current ratchet key, then key advances.
        """
        if len(data) > MAX_MESSAGE_SIZE:
            raise ValueError(f"Message too large: {len(data)} > {MAX_MESSAGE_SIZE}")

        encrypted = self._session.encrypt(data)
        self._write_frame(encrypted)

    def recv(self) -> bytes:
        """
        Receive and decrypt message.

        Verifies authentication and advances receive ratchet.
        """
        encrypted = self._read_frame()
        return self._session.decrypt(encrypted)

    @property
    def service(self) -> str:
        """Get service identifier."""
        return self._service

    @property
    def messages_sent(self) -> int:
        """Get send message count."""
        return self._session.send_counter

    @property
    def messages_received(self) -> int:
        """Get receive message count."""
        return self._session.recv_counter

    def close(self) -> None:
        """Close the channel."""
        self._stream.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    # --- Internal helpers ---

    @staticmethod
    def _compute_session_key(
        config: ChannelConfig,
        salt: bytes,
        local_contribution: bytes,
        remote_contribution: bytes
    ) -> bytes:
        """Compute session key from contributions."""
        # Combine contributions
        base_key = PAKEExchange.combine(local_contribution, remote_contribution)

        # Mix in password-derived secret (not exchanged)
        password_key = PAKEExchange.derive(
            config.password, salt, "session", config.iterations
        )

        # Final session key = hash(base_key || password_key)
        combined = base_key + password_key
        return hashlib.sha256(combined).digest()

    @classmethod
    def _send_handshake(cls, stream: BinaryIO, msg_type: int, data: bytes) -> None:
        """Send handshake message."""
        frame = bytes([PROTOCOL_VERSION, msg_type])
        frame += struct.pack('>H', len(data))
        frame += data
        stream.write(frame)
        stream.flush()

    @classmethod
    def _recv_handshake(cls, stream: BinaryIO, expected_type: int) -> bytes:
        """Receive handshake message."""
        header = stream.read(4)
        if len(header) != 4:
            raise ValueError("Connection closed during handshake")

        version, msg_type, length = header[0], header[1], struct.unpack('>H', header[2:4])[0]

        if version != PROTOCOL_VERSION:
            raise ValueError(f"Unsupported protocol version: {version}")

        if msg_type != expected_type:
            raise ValueError(f"Unexpected message type: expected {expected_type}, got {msg_type}")

        if length > 1024:
            raise ValueError("Handshake message too large")

        data = stream.read(length)
        if len(data) != length:
            raise ValueError("Connection closed during handshake")

        return data

    @classmethod
    def _send_confirmation(cls, stream: BinaryIO, session_key: bytes, is_client: bool) -> None:
        """Send handshake confirmation."""
        label = b"client-confirm" if is_client else b"server-confirm"
        confirm = hmac.new(session_key, label, hashlib.sha256).digest()[:16]
        cls._write_frame_raw(stream, confirm)

    @classmethod
    def _verify_confirmation(cls, stream: BinaryIO, session_key: bytes, expect_client: bool) -> None:
        """Verify handshake confirmation."""
        received = cls._read_frame_raw(stream)
        if len(received) != 16:
            raise ValueError("Invalid confirmation")

        label = b"client-confirm" if expect_client else b"server-confirm"
        expected = hmac.new(session_key, label, hashlib.sha256).digest()[:16]

        if not hmac.compare_digest(received, expected):
            raise ValueError("Authentication failed")

    def _write_frame(self, data: bytes) -> None:
        """Write length-prefixed frame."""
        self._write_frame_raw(self._stream, data)

    def _read_frame(self) -> bytes:
        """Read length-prefixed frame."""
        return self._read_frame_raw(self._stream)

    @staticmethod
    def _write_frame_raw(stream: BinaryIO, data: bytes) -> None:
        """Write length-prefixed frame."""
        stream.write(struct.pack('>I', len(data)))
        stream.write(data)
        stream.flush()

    @staticmethod
    def _read_frame_raw(stream: BinaryIO) -> bytes:
        """Read length-prefixed frame."""
        len_buf = stream.read(4)
        if len(len_buf) != 4:
            raise ValueError("Connection closed")

        length = struct.unpack('>I', len_buf)[0]
        if length > MAX_MESSAGE_SIZE:
            raise ValueError(f"Frame too large: {length} > {MAX_MESSAGE_SIZE}")

        data = stream.read(length)
        if len(data) != length:
            raise ValueError("Connection closed during frame read")

        return data


class ShieldListener:
    """Channel listener for accepting multiple connections."""

    def __init__(self, listener, config: ChannelConfig):
        self._listener = listener
        self._config = config

    def accept(self) -> ShieldChannel:
        """Accept next connection."""
        conn, addr = self._listener.accept()
        return ShieldChannel.accept(conn.makefile('rwb'), self._config)

    @property
    def config(self) -> ChannelConfig:
        """Get configuration."""
        return self._config

    def close(self) -> None:
        """Close listener."""
        self._listener.close()
